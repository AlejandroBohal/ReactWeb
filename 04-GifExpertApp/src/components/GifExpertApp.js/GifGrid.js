import React, { useState,useEffect } from 'react'
import { getGifs } from '../../services/gifAPIClient'
import GifGridItem from '../GifGridItem';

const GifGrid = ({category}) => {
    const [images, setImages] = useState([]);
    useEffect( () =>{
      getGifs('Marvel').then( res => setImages(res));
    },[]);

    return (
        <>
          <h3>{category}</h3>
          <div className="card-grid">
                {
                  images.map( (image) => (
                    <GifGridItem key={image.id} {...image} />
                  ))
                }
          </div>
        </>
    )
}

export default GifGrid
