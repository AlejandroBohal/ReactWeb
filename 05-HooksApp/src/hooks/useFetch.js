import { useEffect, useRef, useState } from "react"

export const useFetch = ( url ) => {
    const [state, setState] = useState({
        data:null,
        loading:true,
        error:null
    });
    //Si el componente es desmontado y la peticiÃ³n se sigue ejecutando esta falla
    //para solucionarlo se utiliza el useRef
    const isMounted = useRef(true);
    useEffect(() => {
        return () => {
            isMounted.current = false; 
        }
    }, []);
    useEffect(() => {
        setState({data:null,error:null,loading:true});
        fetch(url).then(res => res.json()).then(data => {
            if (isMounted.current){
                setState({
                    loading: false,
                    error: null,
                    data
                });
            }
        });
    }, [url]);
    return state;
}
