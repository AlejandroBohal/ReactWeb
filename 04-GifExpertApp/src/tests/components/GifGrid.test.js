import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import React from 'react';
import GifGrid from '../../components/GifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs');

describe('Testing GifGrid', () => {

    const category = 'test category';
    useFetchGifs.mockReturnValue({
        data:[],
        loading:true,
    });
    const wrapper = shallow(<GifGrid category={category}/>);
    test('should load <GifGrid /> correctly', () => {
        
        expect(wrapper).toMatchSnapshot();
    });
    test('should load items when load useFetchGif hook', () => {
        const imgs = [{
            id: 'ABC',
            title: 'GifExample',
            url: 'https://localhost/tests'
        }];
        useFetchGifs.mockReturnValue({
            data: imgs,
            loading:false,
        });
        const wrapper = shallow(<GifGrid category={category}/>);
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('p').exists()).toBe(false);
        expect(wrapper.find('GifGridItem').length).toBe(imgs.length);
    });
    

})
