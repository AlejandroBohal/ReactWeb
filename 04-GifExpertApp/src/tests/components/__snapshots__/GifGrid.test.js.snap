// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing GifGrid should load <GifGrid /> correctly 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    test category
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Loading...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`Testing GifGrid should load items when load useFetchGif hook 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    test category
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="GifExample"
      url="https://localhost/tests"
    />
  </div>
</Fragment>
`;
